outer.c

```diff
@@ -315,6 +324,34 @@ void outer_helper_spawn(struct outer_helper *helper)
 
 	if (helper->cgroup_enabled) {
 		int cgroupfd = recv_fd(fdpair[SOCKET_CHILD]);
+
+		char *subcgroup = makepath("bst.%d", rootpid);
+
+		// Writing to cgroup directory likely requires privilige
+		make_capable(BST_CAP_DAC_OVERRIDE);
+
+		if (mkdirat(cgroupfd, subcgroup, 0777) == -1) {
+			err(1, "unable to create sub-hierachy cgroup %s", subcgroup);
+		}
+
+		reset_capabilities();
+
+		int subcgroupfd = openat(cgroupfd, subcgroup, 0);
+		if (subcgroupfd == -1) {
+			err(1, "unable to open cgroup %s", subcgroup);
+		}
+
+		char srootpid[BUFSIZ];
+		if (sprintf(srootpid, "%d", rootpid) == -1) {
```

Comment:
Is the reason why we can't move all that after child_pid is defined because we enter the cgroup namespace before?

One weakness of the current implementation is that the bst process itself counts towards the cgroup limits, rather than the child process only. It'd be nice if the child was the first process of the new cgroup.

This might require unsharing the cgroups namespace after the initial fork, so maybe not for this pull request. I'm fine if this is done separately, though.

Reply:
Yes that was an issue I encountered in the first draft. We need to child pid to add to the bst cgroup, but this is only accessible after the fork, which is after the ns_enter(). With an extra call to ns_enter() we can re-unshare the cgroups though.

Reply:
My proposal is to unshare the cgroup ns only once, but to do so after the fork. In other words, ns_enter shares enters all namespaces except for the cgroup namespace, which must be done post-fork and post-cgroups-setup